dot/node_test.go

```diff
@@ -0,0 +1,437 @@
+// Copyright 2021 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package dot
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/ChainSafe/gossamer/dot/core"
+	"github.com/ChainSafe/gossamer/dot/digest"
+	"github.com/ChainSafe/gossamer/dot/network"
+	"github.com/ChainSafe/gossamer/dot/state"
+	gsync "github.com/ChainSafe/gossamer/dot/sync"
+	"github.com/ChainSafe/gossamer/dot/system"
+	"github.com/ChainSafe/gossamer/dot/telemetry"
+	"github.com/ChainSafe/gossamer/dot/types"
+	"github.com/ChainSafe/gossamer/internal/log"
+	"github.com/ChainSafe/gossamer/internal/metrics"
+	"github.com/ChainSafe/gossamer/lib/babe"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/genesis"
+	"github.com/ChainSafe/gossamer/lib/grandpa"
+	"github.com/ChainSafe/gossamer/lib/keystore"
+	"github.com/ChainSafe/gossamer/lib/runtime"
+	"github.com/ChainSafe/gossamer/lib/runtime/wasmer"
+	"github.com/ChainSafe/gossamer/lib/services"
+	"github.com/ChainSafe/gossamer/lib/utils"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestInitNode(t *testing.T) {
+	cfg := NewTestConfig(t)
+	genFile := newTestGenesisRawFile(t, cfg)
+	require.NotNil(t, genFile)
+
+	cfg.Init.Genesis = genFile
+
+	type args struct {
+		cfg *Config
+	}
+	tests := []struct {
+		name string
+		args args
+		err  error
+	}{
+		{
+			name: "test config",
+			args: args{cfg: cfg},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := InitNode(tt.args.cfg)
+
+			if tt.err != nil {
+				assert.EqualError(t, err, tt.err.Error())
+			} else {
+				assert.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestLoadGlobalNodeName(t *testing.T) {
+	t.Parallel()
+
+	// initialise database using data directory
+	basePath := t.TempDir()
+	db, err := utils.SetupDatabase(basePath, false)
+	require.NoError(t, err)
+
+	basestate := state.NewBaseState(db)
+	basestate.Put(common.NodeNameKey, []byte(`nodeName`))
+
+	err = db.Close()
+	require.NoError(t, err)
+
+	type args struct {
+		basepath string
+	}
+	tests := []struct {
+		name         string
+		args         args
+		wantNodename string
+		err          error
+	}{
+		{
+			name:         "working example",
+			args:         args{basepath: basePath},
+			wantNodename: "nodeName",
+		},
+		{
+			name: "wrong basepath test",
+			args: args{basepath: "wrong_path"},
+			err:  errors.New("Key not found"),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			gotNodename, err := LoadGlobalNodeName(tt.args.basepath)
+			if tt.err != nil {
+				assert.EqualError(t, err, tt.err.Error())
+				err := os.RemoveAll(tt.args.basepath)
+				require.NoError(t, err)
+			} else {
+				assert.NoError(t, err)
+			}
+
+			assert.Equal(t, tt.wantNodename, gotNodename)
+		})
+	}
+}
+
+func TestNewNode(t *testing.T) {
+	ctrl := gomock.NewController(t)
```

Comment:
See this https://github.com/ChainSafe/gossamer/pull/2239#discussion_r803765992

TLDR: Use dumb subtests with everything locally scoped when you have decently-changing-mocks-configurations-between-test-cases and do not use test case slices + for loop.

Reply:
I removed for loop, looks a little better, but still a lot of mock usage.